### PR TITLE
Classes checking class existence

### LINKED ISSUE #392: slf4j-api-1.7.21.jar (13 errors) Netty
```
slf4j-api-1.7.21.jar (13 errors):
  ClassSymbolReference{sourceClassName=org.slf4j.LoggerFactory, targetClassName=org.slf4j.impl.StaticLoggerBinder, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  ClassSymbolReference{sourceClassName=org.slf4j.MarkerFactory, targetClassName=org.slf4j.impl.StaticMarkerBinder, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  ClassSymbolReference{sourceClassName=org.slf4j.MDC, targetClassName=org.slf4j.impl.StaticMDCBinder, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.LoggerFactory, targetClassName=org.slf4j.impl.StaticLoggerBinder, methodName=getSingleton, interfaceMethod=false, descriptor=()Lorg/slf4j/impl/StaticLoggerBinder;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.LoggerFactory, targetClassName=org.slf4j.impl.StaticLoggerBinder, methodName=getLoggerFactoryClassStr, interfaceMethod=false, descriptor=()Ljava/lang/String;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.LoggerFactory, targetClassName=org.slf4j.impl.StaticLoggerBinder, methodName=getLoggerFactory, interfaceMethod=false, descriptor=()Lorg/slf4j/ILoggerFactory;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.MarkerFactory, targetClassName=org.slf4j.impl.StaticMarkerBinder, methodName=getSingleton, interfaceMethod=false, descriptor=()Lorg/slf4j/impl/StaticMarkerBinder;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.MarkerFactory, targetClassName=org.slf4j.impl.StaticMarkerBinder, methodName=getMarkerFactory, interfaceMethod=false, descriptor=()Lorg/slf4j/IMarkerFactory;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.MDC, targetClassName=org.slf4j.impl.StaticMDCBinder, methodName=getSingleton, interfaceMethod=false, descriptor=()Lorg/slf4j/impl/StaticMDCBinder;}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.slf4j.MDC, targetClassName=org.slf4j.impl.StaticMDCBinder, methodName=getMDCA, interfaceMethod=false, descriptor=()Lorg/slf4j/spi/MDCAdapter;}, reason: CLASS_NOT_FOUND, target class location not found
  FieldSymbolReference{sourceClassName=org.slf4j.LoggerFactory, targetClassName=org.slf4j.impl.StaticLoggerBinder, fieldName=REQUESTED_API_VERSION}, reason: CLASS_NOT_FOUND, target class location not found
  FieldSymbolReference{sourceClassName=org.slf4j.MarkerFactory, targetClassName=org.slf4j.impl.StaticMarkerBinder, fieldName=SINGLETON}, reason: CLASS_NOT_FOUND, target class location not found
  FieldSymbolReference{sourceClassName=org.slf4j.MDC, targetClassName=org.slf4j.impl.StaticMDCBinder, fieldName=SINGLETON}, reason: CLASS_NOT_FOUND, target class location not found
```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #392.

This change is to filter errors on `CLASS_NOT_FOUND` when the source class has a method that catches `NoClassDefFoundError`. For example `org.slf4j.LoggerFactory` has:

```
    private final static void bind() {
        try {
...
            // the next line does the binding
            StaticLoggerBinder.getSingleton();
...
        } catch (NoClassDefFoundError ncde) {
            String msg = ncde.getMessage();
            if (messageContainsOrgSlf4jImplStaticLoggerBinder(msg)) {
                INITIALIZATION_STATE = NOP_FALLBACK_INITIALIZATION;
```

The catch clause is compiled to exception handlers in class file.

# Result

With this change, grpc-netty shows "PASS":

image

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 1]
```diff
@@ -510,6 +510,42 @@ boolean hasValidSuperclass(JavaClass childJavaClass, JavaClass parentJavaClass)
...
+        }
+      }
+    } catch (ClassNotFoundException ex) {
+      // Because the reference in the argument was extracted from the source class file,
+      // the source class should be found.
+      throw new ClassFormatException(
+          "The source class in the reference is no longer available in the class path", ex);
+    }
+
+    // The source class does not have a method that catches NoClassDefFoundERror
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: NoClassDefFoundError

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java [hunk 2]
```diff
@@ -806,4 +806,23 @@ public void testJarPathOrderInResolvingReferences()
...
+    // the tool should not show errors for such classes.
+    List<Path> paths =
+        ClassPathBuilder.artifactsToClasspath(
+            ImmutableList.of(new DefaultArtifact("org.slf4j:slf4j-api:jar:1.7.21")));
+
+    StaticLinkageChecker staticLinkageChecker = StaticLinkageChecker.create(false, paths, paths);
+
+    StaticLinkageCheckReport linkageErrors = staticLinkageChecker.findLinkageErrors();
+
+    JarLinkageReport slf4jLinkageReport = linkageErrors.getJarLinkageReports().get(0);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: we should also have a test for the case where there is no check

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 3]
```diff
@@ -510,6 +510,42 @@ boolean hasValidSuperclass(JavaClass childJavaClass, JavaClass parentJavaClass)
     return constantPoolIndicesForTarget.build();
   }
 
+  /**
+   * Returns true if {@link SymbolReference#getSourceClassName()} has a method that has an exception
+   * handler for {@link NoClassDefFoundError}.
+   */
+  boolean isSourceClassCheckingNoClassDefFoundError(SymbolReference reference) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: perhaps just `catchesNoClassDefFoundError`

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: Renamed.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: Fixed

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: Added test `testFindLinkageErrors_classNotfoundWhenClassesNotCatchingNoClassDefFoundError`

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java [hunk 4]
```diff
@@ -806,4 +806,41 @@ public void testJarPathOrderInResolvingReferences()
...
+    Truth.assertThat(slf4jLinkageReport.getMissingClassErrors()).isEmpty();
+    Truth.assertThat(slf4jLinkageReport.getMissingFieldErrors()).isEmpty();
+    Truth.assertThat(slf4jLinkageReport.getMissingMethodErrors()).isEmpty();
+  }
+
+  @Test
+  public void testFindLinkageErrors_classNotfoundWhenClassesNotCatchingNoClassDefFoundError()
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: testFindLinkageErrors_doesNotCatchNoClassDefFoundError

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java [hunk 5]
```diff
@@ -806,4 +806,41 @@ public void testJarPathOrderInResolvingReferences()
         .that(reportWith66First.getMissingMethodErrors())
         .isEmpty();
   }
+
+  @Test
+  public void testFindLinkageErrors_classesCatchingNoClassDefFoundErrorInSlf4j()
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: testFindLinkageErrors_catchesNoClassDefFoundError

Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: Renamed

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java (hunk 4 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: Renamed

**Pull request merged**